syntax = "proto3";

package calendar;

import "google/protobuf/empty.proto";

// User in system
message UserId { string user_id = 1; }

// OAuth
message OAuthUrl { string url = 1; }
message OAuthTokens {
  string user_id = 1;
  string access_token = 2;
  string refresh_token = 3;
  int64 expiry_epoch = 4;
}

// Calendar event
message Event {
  string id = 1;
  string title = 2;
  string description = 3;
  string start = 4;
  string end = 5;
  string event_type = 6;
  bool is_out_of_office = 7;
}

// CRUD
message CreateEventReq {
  string user_id = 1;
  string calendar_id = 2; // default "primary"
  string title = 3;
  string description = 4;
  string start_iso = 5;
  string end_iso = 6;
  string event_type = 7; // hold | out_of_office | personal | appointment | availability
}
message UpdateEventReq {
  string user_id = 1;
  string calendar_id = 2;
  string event_id = 3;
  string title = 4;
  string description = 5;
  string start_iso = 6;
  string end_iso = 7;
  string event_type = 8;
}
message DeleteEventReq { 
  string user_id = 1; 
  string calendar_id = 2; 
  string event_id = 3; 
}
message ListEventsReq { 
  string user_id = 1; 
  string calendar_id = 2; 
  string time_min = 3; 
  string time_max = 4; 
}

// Wrapper for RPC or API response
message ListEventsResp { 
  repeated Event events = 1; 
}

service CalendarSync {
  // return OAuth consent URL for the user to visit
  rpc GetOAuthUrl(UserId) returns (OAuthUrl);

  // Called after OAuth callback: store tokens
  rpc StoreTokens(OAuthTokens) returns (google.protobuf.Empty);

  // Opt out / revoke sync
  rpc OptOut(UserId) returns (google.protobuf.Empty);

  // CRUD operations
  rpc CreateEvent(CreateEventReq) returns (Event);
  rpc UpdateEvent(UpdateEventReq) returns (Event);
  rpc DeleteEvent(DeleteEventReq) returns (google.protobuf.Empty);
  rpc ListEvents(ListEventsReq) returns (ListEventsResp);

  // For webhook -> server can call
  rpc HandlePushNotification(UserId) returns (google.protobuf.Empty);
}